# mTaxSi

An R package for metabolic exploration of microbial taxonomic signatures.

While large-scale cohort studies have identified long lists of microbes implicated in disease, why they are associated with differential host phenotypes remains unclear. Metabolomics is considered the most promising ‘omic technology to reveal deep mechanistic insights into the gut microbiome by providing a snapshot of host-microbial co-metabolism. However, its use in the field of microbiomics is still in its infancy. Moreover, a key goal of microbiome research is to define a microbe’s function and its impact on the host. We therefore introduce mTaxSi, an R package that enables the metabolic exploration of microbial taxonomic signatures. mTaxSi leverages genome-scale metabolic models to infer differences in metabolism between disease phenotypes based on their taxonomic composition. The results are visualized in a series of plots including bubble chart, heatmap, and Principal Coordinates Analysis plot, each providing a different perspective of metabolic potential between phenotypes. Ultimately, mTaxSi empowers researchers to hypothesize the impact of differences in metabolic function that could lead to their phenotype of interest and prioritize these for future metabolomics experiments. 
